common:
    lid_topic:  "/livox/lidar" 
    imu_topic:  "/livox/imu" 
    con_frame: false # true: if you need to combine several LiDAR frames into one
    con_frame_num: 1 # the number of frames combined
    cut_frame: false # true: if you need to cut one LiDAR frame into several subframes
    cut_frame_time_interval: 0.1 # should be integral fraction of 1 / LiDAR frequency
    time_diff_lidar_to_imu: 0.0 # Time offset between LiDAR and IMU calibrated by other algorithms, e.g., LI-Init (find in Readme)

preprocess:
    lidar_type: 1 # 4
    scan_line: 6 # 32
    timestamp_unit: 1           # the unit of time/t field in the PointCloud2 rostopic: 0-second, 1-milisecond, 2-microsecond, 3-nanosecond.
    blind: 0.5

mapping:
    imu_en: true
    extrinsic_est_en: false # for aggressive motion, set this variable false
    imu_time_inte: 0.005 # = 1 / frequency of IMU
    lidar_time_inte: 0.1
    satu_acc: 3.0 # the saturation value of IMU's acceleration. not related to the units
    satu_gyro: 35 # the saturation value of IMU's angular velocity. not related to the units
    acc_norm: 1.0 # 9.810 # 1.0 for g as unit, 9.81 for m/s^2 as unit of the IMU's acceleration
    lidar_meas_cov: 0.01 # 0.001; 0.01
    acc_cov_output: 500
    gyr_cov_output: 1000 
    b_acc_cov: 0.0001 
    b_gyr_cov: 0.0001 
    imu_meas_acc_cov: 0.1 #0.1 # 0.1
    imu_meas_omg_cov: 0.1 #0.01 # 0.1
    gyr_cov_input: 0.01 # for IMU as input model
    acc_cov_input: 0.1 # for IMU as input model
    plane_thr: 0.1 # 0.05, the threshold for plane criteria, the smaller, the flatter a plane
    match_s: 81
    ivox_grid_resolution: 2.0
    gravity: [0.0, 0.0, -9.810] # [0.0, 9.810, 0.0] # # [0.0, 0.0, -9.787561] # gvins # 
    gravity_init: [0.0, 0.0, -9.810] # preknown gravity in the initial IMU frame for unstationary start or in the initial LiDAR frame for using without IMU
    extrinsic_T: [ 0.011, 0.02329, -0.04412 ] # avia # [0.011, 0.02329, -0.04412] # mid360
    extrinsic_R: [ 1, 0, 0,
                   0, 1, 0,
                   0, 0, 1 ]

odometry: 
    publish_odometry_without_downsample: false

publish:
    path_en: true                 # false: close the path output
    scan_publish_en: true         # false: close all the point cloud output
    scan_bodyframe_pub_en: false  # true: output the point cloud scans in IMU-body-frame

pcd_save:
    pcd_save_en: true
    interval: -1                 # how many LiDAR frames saved in each pcd file; 
                                 # -1 : all frames will be saved in ONE pcd file, may lead to memory crash when having too much frames.

loopClosure:
  ifLoop: false      # 是否开启回环
  ifPCL: true       # pcl或者open3d库
  surroundingkeyframeAddingDistThreshold: 0.1   # meters, regulate keyframe adding threshold
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold  
  historyKeyframeSearchRadius: 5.0              # meters, key frame that is within n meters from current pose will be considerd for loop closure 3.0
  historyKeyframeSearchTimeDiff: 10.0           # seconds, key frame that is n seconds older will be considered for loop closure 30.0 
  historyKeyframeSearchNum: 1                  # 历史帧检测数
  currentKeyframeSearchNum: 30                  # 当前帧检测数
  historyKeyframeFitnessScore: 0.15         # icp threshold, the smaller the better alignment 0.15 
  mappingProcessInterval: 0.15              # 优化时间间隔                
  kdtreeResetTime: 60                       # 重置kd树的最短时间间隔
  kdtreeResetDistance: 1                  # 重置kd树的距离阈值
  loopHeightMax: 0.5                       # ICP配准时最大高度 0.5
  loopHeightMin: -3                        # ICP配准时最小高度 -3        

detectObject: 
    ifFloorDetect: true         # 是否开启楼层检测
    floor_detect_rate: 0.5      # 楼层检测频率
    detect_object_en : true     # 是否开启物体检测
    save_interval: 20           # 每隔多少帧保存一次点云，用于物体检测    
    normal_variance_threshold_deg: 60 
    guess_high: 2.5        